File: Search/binarysearch.cpp
#include<iostream>
#include<math.h>
#include<fstream>
#include<vector>
#include<sstream>
#include<algorithm>

using namespace std;

int counter = 0;

int binarySearch(vector<int> &arr, unsigned int low, unsigned int high, int num){

	if(low > high)
		return -1;
	else
	{
		int mid = (low + high) / 2;
		counter++;

		if(arr[mid] == num)
			return mid;
		else if(arr[mid] < num)
				return binarySearch(arr, mid + 1, high, num);
		else
				return binarySearch(arr, low, mid, num);
		}
}

int main(int argc, char*argv[]){
	int x;
	vector<int> arr;
	int result;
	string line, line2;
	fstream file;
	ifstream infile(argv[1]);
	file.open(argv[2], ios::out);


	getline(infile, line);
	istringstream buffer{line};

	while(buffer >> x){
		arr.push_back(x);

		if(buffer.peek() == ',') // >> std::ws)
			buffer.ignore();
	}
	sort(arr.begin(), arr.end());
	int y;
	while(infile >> y)
	{
		result = binarySearch(arr, 1, arr.size() + 1, y);
		(result == -1) ? file<<to_string(y)+" not found"<<endl : file<<to_string(y)+" found at position "<<result<<" after "<<counter<<" comparisons"<<endl;
	}
	// sort(arr.begin(), arr.end());

}

File: Search/BinarySearch/python/BinarySearch.py
def Binary_search(arr, start_index, last_index, element):
    mid = (int)(start_index + (last_index - start_index)) / 2

    if (start_index > last_index):
        print("Element not found")

    elif (element > arr[mid]):
        start_index = mid + 1
        Binary_search(arr, start_index, last_index, element)

    elif (element < arr[mid]):
        last_index = mid - 1
        Binary_search(arr, start_index, last_index, element)

    else:
        print("element is present at index " + str(mid))


arr = [2, 14, 19, 21, 99, 210, 512, 1028, 4443, 5110]
element = 99
start_index = 0
last_index = len(arr) - 1
Binary_search(arr, start_index, last_index, element)
